client/service/src/chain_ops.rs

```diff
@@ -114,86 +218,100 @@ impl<
 		// This makes it possible either to interleave other operations in-between the block imports,
 		// or to stop the operation completely.
 		let import = future::poll_fn(move |cx| {
-			// Start by reading the number of blocks if not done so already.
-			let count = match count {
-				Some(c) => c,
-				None => {
-					let c: u64 = match Decode::decode(&mut io_reader_input) {
-						Ok(c) => c,
-						Err(err) => {
-							let err = format!("Error reading file: {}", err);
-							return std::task::Poll::Ready(Err(From::from(err)));
-						},
-					};
-					info!("📦 Importing {} blocks", c);
-					count = Some(c);
-					c
-				}
-			};
-
 			// Read blocks from the input.
-			if read_block_count < count {
-				match SignedBlock::<Self::Block>::decode(&mut io_reader_input) {
-					Ok(signed) => {
-						let (header, extrinsics) = signed.block.deconstruct();
-						let hash = header.hash();
-						// import queue handles verification and importing it into the client
-						queue.import_blocks(BlockOrigin::File, vec![
-							IncomingBlock::<Self::Block> {
-								hash,
-								header: Some(header),
-								body: Some(extrinsics),
-								justification: signed.justification,
-								origin: None,
-								allow_missing_state: false,
-								import_existing: force,
-							}
-						]);
-					}
-					Err(e) => {
-						warn!("Error reading block data at {}: {}", read_block_count, e);
-						return std::task::Poll::Ready(Ok(()));
+			match Pin::new(&mut block_stream).poll_next(cx) {
+				Poll::Ready(None) => {
+					let read_block_count = block_stream.read_block_count();
+					if let Some(count) = block_stream.count() {
+						if link.imported_blocks >= count {
+							info!(
+								"🎉 Imported {} blocks. Best: #{}",
+								count, client.chain_info().best_number
+							);
+							// We're done importing blocks, we can stop here.
+							return std::task::Poll::Ready(Ok(()))
+						} else {
+							return std::task::Poll::Pending
+						}
+					} else {
+						if link.imported_blocks >= read_block_count {
```

Comment:
I'm not sure this condition is correct. Initially we would check `link.imported_blocks >= count` and not `read_block_count` however `count` is not available in JSON format...